lib/composite_primary_keys/associations/collection_association.rb

```diff
@@ -2,17 +2,24 @@ module CompositePrimaryKeys
   module CollectionAssociation
     def ids_writer(ids)
       primary_key = reflection.association_primary_key
-      pk_type = klass.type_for_attribute(primary_key)
       ids = Array(ids).reject(&:blank?)
-      ids.map! { |i| pk_type.cast(i) }
 
       # CPK-
       if primary_key.is_a?(Array)
+        ids = CompositePrimaryKeys.normalize(ids, primary_key.size)
+        primary_key.each_with_index do |primary_key, i|
```

Comment:
Reusing the variable primary_key in the block is confusing - it should be named something else.